### PR TITLE
GHA: update build matrix

### PR DESCRIPTION
**Pull request by dyfer**
## Purpose and Motivation

~~Draft PR to explore issues with the toolchain upgrades~~

This PR updates the build matrix:
- Linux
  - enable clang 16-18 builds
- macOS:
  - separate arm64 and x64 builds
  - remove `brew-install-universal` script and the "universal" build, which took a long time
  - use qt6 from aqt install to provide compatibility with macOS versions lower than the build OS (I checked, the builds using homebrew qt@6 done on macOS 13 did not run on macOS 12)
    - update qt to version 6.7.3, the last one compatible with macOS 11
  - change structure of the build script, grouping installation steps by package source, rather than package name
  - minor cleanup of the yaml file - cache keys etc
- Windows:
  - add MSVC 2022 build and use artifacts from it for releases etc
  - update qt to version 6.7.3
  - stop uploading windows 32-bit builds
  - minor cleanup of the yaml file - cache keys etc

Outstanding issues:
- on macOS qt@6 from aqt install takes a relatively long time to install when not cached and the cache is large; however cached install is not too bad. We can't do anything about this anyway...
- on macOS both x64 and arm64 builds are large; the qt libraries in the app bundle end up including both architectures (i.e. are "universal"). My understanding is that `macdeployqt` [should "strip" the binaries by default](https://doc.qt.io/qt-6/macos-deployment.html#:~:text=dmg%20disk%20image-,%2Dno%2Dstrip,-Don%27t%20run%20%27strip), leaving just one architecture, but this doesn't seem to be the case here.
  - this nullifies the "smaller binaries" argument of dropping the universal build, but the separate builds are still a bit faster

## Types of changes

- New feature / maintenance
- Documentation

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Comment by capital-G** [comment]
Do you think we should/could use this to make the CI a less "branchy" and stateful for macOS or should this be done later?

I looked today into separating the macOS build files to separate files and I found it much clearer to understand what is happening, and also more confident to modify it if necessary. The downside is ofc that there is some overlap in their bootstrap code: https://github.com/capital-G/supercollider/blob/7055fd46e614ae970719d938d2dd34fe38ccd7b0/.github/workflows/build_macos_universal.yml

What I learned from it
* having multiple action files seems a bit clumsy in GHA as there is no way to bundle builds beyond using matrix
* using a "cache-prefix" for each build makes much easier to follow what is happening

I think we could make the heavy lifting of setting up the environment through shell scripts which are not nested state rather clearly what they are doing?

A matrix could look like

```
- name
- cache-prefix
- shell-script-to-setup-dependencies
- CMAKE-flags 
```

with steps

```
- setup cache
- invoke setup shell script
- configure
- build
- bundling for artifacts 
```

On the other hand: Do not change a running system? Should we take a closer look at this or do you think it's fine "as is"?

**Comment by dyfer** [comment]
I wouldn't go towards separate files. 

I'm not sure what "nested state" means, could you clarify?

We could certainly consider reorganizing steps; e.g. instead of separating homebrew qt5 install, do all homebrew in one step, all vcpkg in one step etc.

Aside from using various package tools (homebrew vs vcpkg) for different libraries, the build steps are rather clear IMO:
- setup
- cache
- library install
- configure
- build
- upload artifact

You also mentioned using shell scripts - do you mean scripts outside of the yml files? This is how Travis CI was set up back in the day and I personally found it really difficult to track down which sub-script was doing what at which moment. I find the current order of operation quite clear - but also likely because I wrote a good chunk of it.

**Comment by capital-G** [comment]
E.g.

```shell
      - name: install dependencies
        run: |
          brew install ccache
          # add ccamke to PATH - see https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions
          echo "/usr/local/opt/ccache/libexec" >> $GITHUB_PATH
          if [[ "${{ matrix.system-portaudio }}" != "false" ]]; then brew install portaudio; fi
          if [[ "${{ matrix.build-fftw }}" != "true" ]]; then brew install fftw; fi

          # install universal versions of homebrew libraries
          if [[ "${{ matrix.cmake-architectures }}" != "x86_64" ]]; then
              UNIVERSAL_PACKAGES=
              if [[ "${{ matrix.build-libsndfile }}" != "true" ]]; then UNIVERSAL_PACKAGES="libsndfile $UNIVERSAL_PACKAGES"; fi
              if [[ "${{ matrix.build-readline }}" != "true" ]]; then UNIVERSAL_PACKAGES="readline $UNIVERSAL_PACKAGES"; fi
              if [[ "${{ matrix.build-fftw }}" != "true" ]]; then UNIVERSAL_PACKAGES="fftw $UNIVERSAL_PACKAGES"; fi
              if [[ "${{ matrix.system-portaudio }}" != "false" ]]; then UNIVERSAL_PACKAGES="portaudio $UNIVERSAL_PACKAGES"; fi
              echo "Downloading script for creating universal binaries from homebrew packages"
              curl --no-progress-meter -O -J -L https://gist.githubusercontent.com/dyfer/6c83905d4593750105897e51e87ec345/raw/98e073d755cc3ca8426e20c30f51e9bac2bc9858/brew-install-universal.sh
              chmod +x brew-install-universal.sh
              ./brew-install-universal.sh $UNIVERSAL_PACKAGES
```

works but also gets really nested. If this would be the only step it would be OK, but there are many if conditions which intervene across steps that make it hard to follow everything and give a better understanding if a change on one step could have side effects in an other build.

Maybe the shell script idea is indeed not a good idea, but maybe we can make the pipeline a bit less dynamic and more static? E.g. something along

```yml
      matrix:
        include:
          - job-name: "universal"
            os-version: "12"
            xcode-version: "13.4.1"
            deployment-target: "10.15"
            cmake-architectures: "x86_64;arm64"
            brew-installs: "ccache"
            brew-universal-installs: "qt@5 fftw libsndfile readfile"
            vcpkg-installs: ""
```

would be much clearer instead of looking up how `system-portaudio` is interpreted within the context of the current build.

The steps would then be

- git clone
- setup cache
- install brew
- install brew universal
- install vcpkg
- configure
- build
- (bundle artifact)

**Comment by dyfer** [comment]
Got it!

I see what you mean about the script. But the script is going away, I think, check out this PR.

I see how 
```
            brew-installs: "qt@5 fftw libsndfile readfile"
            vcpkg-installs: ""
            cmake-flags: ...
```
would be clearer though. I could go both ways with cmake flags being manual or automated, but a list of packages for each environment would definitely be more readable.

**Comment by capital-G** [comment]
> Eg Qt5 prefix path, this should be taken from `brew --prefix` as opposed to being hardcoded (note how this PR replaces hardcoded path to ccache with `brew --prefix`) - I believe taking advantage of automation there is beneficial. IMO it's better to set some cmake flags automatically.
> OTOH I agree that `-DSC_VERIFY_APP=ON` and `-DLIBSCSYNTH=ON` can absolutely be specified manually.

I think it is good to have the CI as explicit as possible and avoid any kind of state/mutation/sideffecte due to the combination of not having the ability to simulate a run locally and waiting multiple minutes for every try.

And while something along `brew --prefix qt@5` would be nice, I think `/opt/homebrew/opt/qt@5 ` is just as valid in the end. If it breaks, we would get aware of it.

**Comment by dyfer** [comment]
@capital-G I updated this PR and it's ready for review

File: .github/workflows/build_macos.yml [hunk 1]
```diff
@@ -184,30 +165,22 @@ jobs:
...
-
-          if [[ "${{ matrix.verify-app }}" == "true" ]]; then EXTRA_CMAKE_FLAGS="-DSC_VERIFY_APP=ON $EXTRA_CMAKE_FLAGS"; fi
-
-          if [[ "${{ matrix.system-portaudio }}" == "false" ]]; then EXTRA_CMAKE_FLAGS="-DSYSTEM_PORTAUDIO=OFF $EXTRA_CMAKE_FLAGS"; fi
+          CMAKE_FLAGS="-G Xcode -D RULE_LAUNCH_COMPILE=ccache -D SUPERNOVA=ON  ${{ matrix.extra-cmake-flags }}"
 
           if [[ -n "${{ matrix.vcpkg-triplet }}" ]]; then
             export VCPKG_ROOT=$VCPKG_INSTALLATION_ROOT
-            EXTRA_CMAKE_FLAGS="-DVCPKG_TARGET_TRIPLET=${{ matrix.vcpkg-triplet }} -DCMAKE_BUILD_TYPE=Release $EXTRA_CMAKE_FLAGS"
+            CMAKE_FLAGS="-DVCPKG_TARGET_TRIPLET=${{ matrix.vcpkg-triplet }} -DCMAKE_BUILD_TYPE=Release $CMAKE_FLAGS"
```
[the remark this answers is not in the document]
Comment (dyfer) [reply .github/workflows/build_macos.yml:178]: In that case you'd need to specify vcpkg triplet twice in the matrix, once in cmake flags, and once for the vcpkg. I think that would be suboptimal... do you think otherwise?

File: .github/workflows/actions.yml [hunk 2]
```diff
@@ -87,7 +87,7 @@ jobs:
     needs: [version-getter, build-macos]
     uses: ./.github/workflows/test_macos.yml
     with:
-      artifact-file: "SuperCollider-${{ needs.version-getter.outputs.sc-version }}-macOS-universal.dmg"
+      artifact-file: "SuperCollider-${{ needs.version-getter.outputs.sc-version }}-macOS-x64.dmg"
```
[the remark this answers is not in the document]
Comment (dyfer) [reply .github/workflows/actions.yml:90]: I wouldn't say intel is a legacy platform just yet... but I do get your point.
In short, the practical reason for this is that if we run tests on macOS >12, we encounter #6488

File: .github/workflows/build_macos.yml [hunk 3]
```diff
@@ -13,21 +13,31 @@ jobs:
       fail-fast: false
       matrix:
         include:
-          - job-name: "universal"
+          - job-name: "arm64"
+            os-version: "14"
```
[the remark this answers is not in the document]
Comment (dyfer) [reply .github/workflows/build_macos.yml:17]: 15 GHA image is still in the beta I think and it just came out. I think we can wait a little before using that, but I could go either way here.

File: .github/workflows/build_macos.yml [hunk 4]
```diff
@@ -123,57 +128,33 @@ jobs:
...
+          # vcpkg is not automatically installed in newer macOS runner images
+          if [[ -n "${{ matrix.vcpkg-triplet }}" && -z "${VCPKG_INSTALLATION_ROOT}" ]]; then
+            cd ${{ github.workspace }}/..
+            echo "setting up vcpkg in `pwd`"
+            git clone https://github.com/microsoft/vcpkg.git
+            cd vcpkg
+            ls -a
+            ./bootstrap-vcpkg.sh
+            echo "VCPKG_INSTALLATION_ROOT=`pwd`" >> $GITHUB_ENV # use the same env var that was used for VCPKG in older runner images
+            echo "`pwd`" >> $GITHUB_PATH
```
[the remark this answers is not in the document]
Comment (dyfer) [reply .github/workflows/build_macos.yml:146]: Sure, I'll change this.

File: .github/workflows/build_windows.yml [hunk 5]
```diff
@@ -21,32 +21,43 @@ jobs:
             qt-arch: "win32_msvc2019"
             fftw-url: "ftp://ftp.fftw.org/pub/fftw/fftw-3.3.5-dll32.zip"
             cmake-generator: "Visual Studio 16 2019"
-            msvc-year: "2019"
-            vcvars-script: "vcvars32.bat"
+            vcvars-script-path: "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Auxiliary/Build/vcvars32.bat"
```
[the remark this answers is not in the document]
Comment (dyfer) [reply .github/workflows/build_windows.yml:24]: I don't think we need 32-bit windows build, but this change would partially answer that question. If the need comes up, we can simply re-enable artifact uploads.
I think the main use for this is to test building SC on a 32-bit platform (not Windows specifically). But maybe indeed this is not needed anymore.
I'd prefer to keep it for now and try to gauge (maybe as on the forum) the need for any 32-bit builds.

File: README.md [hunk 6]
```diff
@@ -26,8 +26,8 @@ See the [Raspberry Pi](README_RASPBERRY_PI.md) and [BeagleBone Black](README_BEA
 ### Platform support
 
 SuperCollider is tested with:
-- Windows 10 64-bit and MSVC 2019
-- macOS 13 and Xcode 14.1
+- Windows 10 64-bit and MSVC 2022
+- macOS 12 and Xcode 15.2
```
[the remark this answers is not in the document]
Comment (dyfer) [reply README.md:30]: We run tests on macOS 12, because of https://github.com/supercollider/supercollider/issues/6488, that's why I made this change.

**Comment by dyfer** [comment]
> (but also lacks this awesome shell script of yours hosted via gist :D)

Thanks! The script might be awesome, but ultimately it's a hack, so it's best not to use it :)

> * IIUC we are currently running the tests on the Ubuntu 22.04 build, maybe we can use the 24.04 build for testing while we are at it?

When I tried running tests on 24.04, I encountered issues. I can't remember what they were, from the top of my head. I'd leave this for another time, as it might need to be investigated further.

> * What are the following macOS builds used for?
>   
>   * x64 shared libscsynth
>   * x64 use system libraries (IIRC this is b/c of boost/flucoma?)
> 
>   Maybe we can add a small comment for motivation b/c I only come across these builds in the gha pipeline and if they start to fail it would be good to have some background for a triage.

Both options test particular cmake build configurations: in the first case `LIBSCSYNTH`, in the second case `SYSTEM_BOOST` and `SYSTEM_YAMLCPP`, with boost and yaml provided by a package manager. Also, this is unrelated to flucoma.

I think the rationale for adding these configurations back in the day was that they are the supported build options and should work on any platform (partial Windows support notwithstanding), so we test whether that's actually the case.

I think that if we want to consider removing these, that should be a separate discussion.


> * Do we keep the w32 build?

I replied in the comment.

File: .github/workflows/actions.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (capital-G) [reply .github/workflows/actions.yml:90]: Oh right, this one. Then lets keep it this way.

I sadly think intel is a legacy platform (still using one myself) :/ We should focus more towards arm because the quantity of intel mac users will not increase in the future.

File: .github/workflows/build_macos.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (capital-G) [reply .github/workflows/build_macos.yml:17]: I don't think its in beta, see https://github.com/github/roadmap/issues/986
I already updated to macOS 15, works w/o any problems so far.

File: .github/workflows/build_macos.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (capital-G) [reply .github/workflows/build_macos.yml:178]: I personally think this provides a more explicit and clearer way and as both entries are not too far apart and can be found by searching for `vcpkg`, I think it can be justified.

File: .github/workflows/build_windows.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (capital-G) [reply .github/workflows/build_windows.yml:24]: I thought that we may could use linux for a 32bit build, but a quick google search reveals that even debian is removing 32bit support for x64_86.
I think arm32 bit seems like a more relevant build if we want to have a 32 bit build?

In short, in this case it makes sense to keep it, but could be re-evaluated once we have an arm pipeline.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (capital-G) [reply README.md:30]: Oh - sorry, you are right.

File: .github/workflows/build_macos.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dyfer) [reply .github/workflows/build_macos.yml:17]: Screenshot 2024-10-13 at 3 15 03 PM
I think it still is in public beta, 
But I think we can still use it, especially now that we are not using qt from homebrew (which tends to bump up the minimum deployment target)

File: .github/workflows/build_macos.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dyfer) [reply .github/workflows/build_macos.yml:178]: While I prefer to not type this twice. We also don't really encourage users to use vcpkg on macOS, we do that "internally" for releases. I'd prefer to keep this hidden. 
BUT if you insist, I'll change it.

**Comment by capital-G** [comment]
> When I tried running tests on 24.04, I encountered issues. I can't remember what they were, from the top of my head. I'd leave this for another time, as it might need to be investigated further.

Sounds reasonable.

> I think the rationale for adding these configurations back in the day was that they are the supported build options and should work on any platform (partial Windows support notwithstanding), so we test whether that's actually the case.

In this case we could consider moving them to linux as the linux builds are much faster. But this would be something for another PR/day - thanks for sharing your insights.

**Comment by dyfer** [comment]
> In this case we could consider moving them to linux as the linux builds are much faster. But this would be something for another PR/day - thanks for sharing your insights.

They already are present in linux. The point was to test whether they work on macOS too.
I agree that this is for another PR/day.

File: .github/workflows/build_macos.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dyfer) [reply .github/workflows/build_macos.yml]: I switched arm64 job to macOS 15, seems to work ok.

File: .github/workflows/build_macos.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (dyfer) [reply .github/workflows/build_macos.yml]: I moved vcpkg to a separate step.

File: .github/workflows/build_macos.yml [hunk 7]
```diff
@@ -123,57 +128,35 @@ jobs:
...
+          echo "`brew --prefix ccache`/libexec" >> $GITHUB_PATH
 
-      - name: install libsndfile # to make it compatible with older OSes (lower deployment target)
-        if: matrix.build-libsndfile == true && matrix.vcpkg-triplet
+      - name: setup vcpkg # vcpkg is not automatically installed in newer macOS runner images
+        if: matrix.vcpkg-packages && matrix.vcpkg-triplet
         run: |
-          brew unlink libsndfile libvorbis libogg flac opus || true
-          vcpkg install libsndfile --triplet="${{ matrix.vcpkg-triplet }}" --overlay-triplets="$GITHUB_WORKSPACE/vcpkg/triplets"
+          if [[ -z "${VCPKG_INSTALLATION_ROOT}" ]]; then
```
Comment (capital-G) [inline .github/workflows/build_macos.yml:140]: why don't we move the if condition in the shell script into the if condition of the step?

File: .github/workflows/build_macos.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (capital-G) [reply .github/workflows/build_macos.yml:180]: Nah, I don't insist, it is merely a matter of taste from my side. As you work currently most with vcpkg I think you can decide upon this, so lets keep it this way.

**Review by capital-G — CHANGES_REQUESTED** [review]
One thing left

File: .github/workflows/build_macos.yml (hunk 7 above)
Comment (dyfer) [reply .github/workflows/build_macos.yml:140]: Because the build matrix doesn't have access to environment variables of the runner. I thought of this, but alas that doesn't work.

Comment (capital-G) [reply .github/workflows/build_macos.yml:140]: There seems something like this possible https://stackoverflow.com/a/74386472 but it is questionable if this is easier to follow then the shell if condition.
Let's keep it this way then :)

**Review by capital-G — APPROVED** [review]

**Comment by capital-G** [comment]
Feel free to merge @dyfer - I don't think there currently are too many people around who are into CI, so I think one review must be sufficient here.

File: .github/workflows/build_macos.yml (hunk 7 above)
Comment (dyfer) [reply .github/workflows/build_macos.yml:140]: Yeah, absolutely it's possible to "export" that value from a previous step, but that definitely introduces more noise than this is worth. Direct "if <envVar>" would be great, but when that's not possible, what we have here is clean enough I think.
